aws/signer/v4/handler.go

```diff
@@ -0,0 +1,146 @@
+	} else {
+		return fmt.Sprintf("%s: %s", msg, e.cause)
+	}
+}
+
+// Unwrap returns the underlying error if one is set
+func (e *HashComputationError) Unwrap() error {
+	return e.err
+}
+
+// SigningError indicates an error condition occurred while performing SigV4 signing
+type SigningError struct {
+	cause string
+	err   error
+}
+
+func (e *SigningError) Error() string {
+	msg := "failed to sign request"
+	if len(e.cause) == 0 {
+		return msg
+	} else {
+		return fmt.Sprintf("%s: %s", msg, e.cause)
+	}
+}
+
+// Unwrap returns the underlying error cause
+func (e *SigningError) Unwrap() error {
+	return e.err
+}
+
+type payloadHashKey struct{}
+
+// UnsignedPayloadMiddleware sets the SigV4 request payload hash to unsigned
+type UnsignedPayloadMiddleware struct{}
+
+func (m *UnsignedPayloadMiddleware) Name() string {
+	return "SigV4 unsigned payload middleware"
+}
+
+func (m *UnsignedPayloadMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (out middleware.FinalizeOutput, err error) {
+	ctx = context.WithValue(ctx, payloadHashKey{}, v4Internal.UnsignedPayload)
+
+	return next.HandleFinalize(ctx, in)
+}
+
+// ComputePayloadHashMiddleware computes sha256 payload hash to sign
+type ComputePayloadHashMiddleware struct{}
+
+// Name is the middleware name
+func (m *ComputePayloadHashMiddleware) Name() string {
+	return "SignV4 Payload Hash Middleware"
+}
+
+// HandleFinalize compute the payload has fof the request payload for SigV4 authentication
+func (m *ComputePayloadHashMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (out middleware.FinalizeOutput, err error) {
+	req := in.Request.(*smithyHTTP.Request)
+
+	body, ok := req.Stream.(io.ReadSeeker)
+	if !ok {
+		return middleware.FinalizeOutput{}, &HashComputationError{cause: errPayloadNotSeekable}
+	}
+
+	_, err = body.Seek(0, io.SeekStart)
+	if err != nil {
+		return middleware.FinalizeOutput{}, &HashComputationError{err: err}
+	}
+
+	hash := sha256.New()
+	_, err = io.Copy(hash, body)
+	if err != nil {
+		return middleware.FinalizeOutput{}, &HashComputationError{err: err}
+	}
+
+	_, err = body.Seek(0, io.SeekStart)
+	if err != nil {
+		return middleware.FinalizeOutput{}, &HashComputationError{err: err}
+	}
+
+	ctx = context.WithValue(ctx, payloadHashKey{}, hex.EncodeToString(hash.Sum(nil)))
+
+	return next.HandleFinalize(ctx, in)
+}
+
+// SignHTTPRequestMiddleware is a `FinalizeMiddleware` implementation for SigV4 HTTP Signing
+type SignHTTPRequestMiddleware struct {
+	Signer interface {
+		SignHTTP(ctx context.Context, r *http.Request, payloadHash string, service string, region string, signingTime time.Time)
+	}
+}
+
+func NewSignHTTPRequestMiddleware(signer interface {
+	SignHTTP(ctx context.Context, r *http.Request, payloadHash string, service string, region string, signingTime time.Time)
+}) *SignHTTPRequestMiddleware {
+	return &SignHTTPRequestMiddleware{Signer: signer}
```

Comment:
This pkg could define its own `HTTPSigner` interface type for the `SignHTTP` interface needed. Would help with being able to document this parameter type potentially.